### ISSUE TITLE
Can I use tensorflow on Windows10 with c++ and gpu support

### ISSUE THREAD

**Issue by xiaoguai0992** [tensorflow/tensorflow#8773]
I wish to use tensorflow on windows10 with c++ and gpu support. Is there any pre build SDK for windows? Or I need to compile from sources?

when I tried to compile on windows with cmake and MSVC, smaple trainer cannot build correctly because of some header files missing, such as "graph.pb.h". Is there any solutions?

ENV:
windows10 vs2015 cmake3.6.2

**Comment by gunan** [comment]
This information is available in our website.
Please make sure you search for solutions before filing issues:
https://www.tensorflow.org/install/install_windows

**Comment by sohaib-l** [comment]
@gunan Following the link you shared it shows only how to install Tensorflow for Python. However, the question of @xiaoguai0992  was about installing Tensorflow on Windows 10 with **C++**, where on the website, nothing was mentionned regarding this issue. 
Tensorflow is an excellent framework and no body can say the opposit. However, it is mainly focused on python + Linux/Mac. I am a windows user, I use a lot Tensorflow/Keras with python but I'd love to see a demo on how to install Tensorflow with **C++** on **WINDOWS** properly. I know it may not be a priority right now, but a lot of people are learning tensorflow, making their own models, but for deployment, we may want to get the best of it, and c/c++ are better than python (I guess ^^). We may also have a lot of projects using c++ and we want to use tensorflow in these projects (I do).

**Comment by gunan** [comment]
Sorry for the misunderstanding.
Then, our instructions would be here:
https://www.tensorflow.org/install/install_c

I see that these do not mention windows, but I think we also build c libraries for windows.
@asimshankar how different are the instructions for windows?

**Comment by asimshankar** [comment]
We do build C libraries for Windows, but only with CPU support. Building GPU-enabled binaries as part of the release is on my TODO list but I haven't gotten around to it yet.

- For CPU, you can download the `.dll` from: https://storage.googleapis.com/tensorflow/libtensorflow/libtensorflow-cpu-windows-x86_64-1.4.0.zip

- We use bazel to build this, but the environment is a bit complicated. The script we use (which is what is run by the [nightly builds](https://ci.tensorflow.org/view/Nightly/job/nightly-libtensorflow-windows/) as well) is [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/windows/libtensorflow_cpu.sh)

So, that sort of summarizes the state we're in with regards to the C API and Windows (and I believe other language bindings like [those for C#](https://github.com/migueldeicaza/TensorFlowSharp) are using these releases) 
I'm happy to provide some pointers to next steps, but we (TensorFlow maintainers) are probably not going to have the bandwidth to do this ourselves in the near future.

- If you really need the C++ API (as opposed to the smaller C API), then a similar process should work, except that you'll use bazel to build the `//tensorflow:libtensorflow_cc.so` target instead of `//tensorflow:libtensorflow.so` target.

- For GPU support: The script outlined about should work, as long as the build was configured to use CUDA. But I have not tried that yet.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by nikhilk** [comment]
+1 to getting this to be officially listed in the documentation along with GPUs of course. :)

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by jackyko1991** [comment]
I just have a look at https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/cmake/tf_shared_lib.cmake

this file seems use a very tricky way to generate dll using cmake.

If you directly follow the instruction in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/cmake/README.md

The tensorflow_BUILD_SHARED_LIB option is default to be off, I am testing on building shared lib with this option

Also this cmake file can generate TensorflowConfig.cmake after INSTALL build, location will be $INSTALL_DIR/lib/cmake, not sure if everything is alright, but worth a test

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 64 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 79 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by andreas-eberle** [comment]
What's the status on this? It would be great to have GPU support in the C/C++ API!

**Comment by jackyko1991** [comment]
@andreas-eberle I have a C++ API example running on window with GPU [here](https://github.com/jackyko1991/vnet-tensorflow/tree/master/cxx)

I have tried to issue pull request on tensorflow main repository but they never merge my code

**Comment by andreas-eberle** [comment]
@jackyko1991: Can you link your PR?

**Comment by jackyko1991** [comment]
@andreas-eberle what PR refers to?

**Comment by andreas-eberle** [comment]
Pull Request. I mean the Pull Request you mentioned that didn't get merged.

**Comment by jackyko1991** [comment]
https://github.com/tensorflow/tensorflow/pull/16480

I have updated the pull request for several times. It is hard for me to catch up with lastest tf release for their slow response...I have stopped to modify the code sinice 1.8.0

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by case540** [comment]
I was under the impression that our Windows GPU release builds were broken (due to some protobuf issue).

**Comment by gunan** [comment]
Protobuf issue is fixed. As far as I can remember, we cherrypicked that fix into 1.9. But I am not 100% on that.
1.10 definitely has that fix.

**Comment by dimakarp1996** [comment]
@asimshankar   Thanks for your explanation, but I still have problems building Tensorflow C++ library for Windows 10, even for CPU. I have replaced tensorflow_jni.so with tensorflow_jni.dylib in libtensorflow_cpu.sh and in BUILD file. I have also replaced libtensorflow.so with libtensorflow.dll and libtensorflow_cc.so with libtensorflow_cc.dll in these two files, as advised in one of the comments. Otherwise, I had error like this
"/tensorflow/tensorflow/tools/lib_package/BUILD:153:1: no such package '@png_archive//'" even though I have installed GNUWin32 which should also contain this package.
Still, the problem persists, as now it complains on the absense of package  '@local_config_git//' .  What should I do to fix it?

**Comment by Cory-Kramer** [comment]
What is the current status of building a tensorflow C (or C++) dll on windows?
From what I've been reading here and elsewhere, the CMake build process is deprecated.
The current Bazel instructions only include C instructions for Mac or Linux, not Windows.
Is there any documentation to use Bazel (or whatever the preferred build process is) for building on Windows?

**Comment by gunan** [comment]
@meteorcloudy

**Comment by jackyko1991** [comment]
if possible can you provide the method for library linking and header file locations?

**Comment by meteorcloudy** [comment]
Working on https://github.com/tensorflow/tensorflow/pull/24963

**Comment by rtrahms** [comment]
Any update on this?  Still wondering if TF Win10 C++ and GPU support will happen.

**Comment by TsarF** [comment]
So, any updates @gunan because I think, that there is a pretty large audience that wants to use c++ with Win10, including me.

**Comment by gunan** [comment]
looks like the referenced PR is merged, and it is definitely in releases.
@meteorcloudy did we have any other issues remaining for this?

**Comment by meteorcloudy** [comment]
@TsarF Does https://github.com/tensorflow/tensorflow/pull/24963 solve your problem?

**Comment by mohantym** [comment]
Hi @xiaoguai0992 !
We are checking to see whether you still need help in this issue . Have you checked this[ thread](https://iq.opengenus.org/build-tensorflow-cpp-library/) on building Tensorflow with C++ using Bazel yet? Thanks!

**Comment by sushreebarsa** [comment]
@xiaoguai0992 Could you please have a look on the [Build from source ](https://www.tensorflow.org/install/source_windows)and let us know if it helps?Thanks!

**Comment by sullivak-kyra** [comment]
@sushreebarsa do those instructions allow for c++ support? It looks like it may only support building a pip package.

**Comment by xiaoguai0992** [comment]
Thank all of your comments! I have already turned to python. If someday I need to play with C++ and Windows again, I'd like to try your suggestions.

**Comment by sushreebarsa** [comment]
@xiaoguai0992 Thank you for the update! Could you please move this issue to closed status if it resolved for you?
Thanks!

**Comment by sullivak-kyra** [comment]
The original question specifically mentioned c++. @xiaoguai0992 moving to python doesn't really close this issue.

**Comment by jackyko1991** [comment]
there are two legacy repo to use tensorflow with C++

https://github.com/cjweeks/tensorflow-cmake
https://github.com/ksachdeva/tensorflow-cc-examples

Seems they are using bazel build, not sure if they can still be used in 2022 release

**Comment by tilakrayal** [comment]
Can you please check this SO [link](https://stackoverflow.com/questions/33620794/how-to-build-and-use-google-tensorflow-c-api) with the similar issue.Thanks
